pageserver/src/layered_repository.rs

```diff
@@ -1422,10 +1456,22 @@ impl LayeredTimeline {
             // Finally, replace the frozen in-memory layer with the new on-disk layers
             layers.remove_historic(frozen.clone());
 
+            let mut historical_layers = Vec::<Arc<dyn Layer>>::with_capacity(
+                new_historics.delta_layers.len() + new_historics.image_layers.len(),
+            );
+            for delta_layer in new_historics.delta_layers {
+                layer_uploads.push(delta_layer.path());
+                historical_layers.push(Arc::new(delta_layer));
+            }
+            for image_layer in new_historics.image_layers {
+                layer_uploads.push(image_layer.path());
+                historical_layers.push(Arc::new(image_layer));
+            }
+
             // If we created a successor InMemoryLayer, its predecessor is
             // currently the frozen layer. We need to update the predecessor
             // to be the latest on-disk layer.
-            if let Some(last_historic) = new_historics.last() {
+            if let Some(last_historic) = historical_layers.last() {
```

Comment:
There's an important meaning to the last historical layer: it is the new predecessor.
This implementation gets it right, by doing the image_layer loop after the delta_layer one. It seems kind of brittle to me